pkg/cloud/azure/actuators/machineset/controller.go

```diff
@@ -132,30 +125,99 @@ func isInvalidConfigurationError(err error) bool {
 }
 
 func (r *Reconciler) reconcile(machineSet *machinev1.MachineSet) (ctrl.Result, error) {
-	providerConfig, err := getproviderConfig(machineSet)
+	klog.Infof("%v: Reconciling MachineSet", machineSet.Name)
+	sku, err := getSKU(r, machineSet)
 	if err != nil {
-		return ctrl.Result{}, mapierrors.InvalidMachineConfiguration("failed to get providerConfig: %v", err)
-	}
-	instanceType, ok := InstanceTypes[strings.ToLower(providerConfig.VMSize)]
-	if !ok {
-		klog.Errorf("Unable to set scale from zero annotations: unknown instance type: %s", providerConfig.VMSize)
-		klog.Errorf("Autoscaling from zero will not work. To fix this, manually populate machine annotations for your instance type: %v", []string{cpuKey, memoryKey, gpuKey})
+		if strings.Contains(err.Error(), "not found in location") {
+			// Print different error message when there is no failure, but SKU is not available.
+			klog.Errorf("Unable to set scale from zero annotations: instance type unknown or unavailabe for this account or location: %v", err)
+		} else {
+			klog.Errorf("Unable to set scale from zero annotations: Azure list SKU request failed: %v", err)
+		}
 
 		// Returning no error to prevent further reconciliation, as user intervention is now required but emit an informational event
-		r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "FailedUpdate", "Failed to set autoscaling from zero annotations, instance type unknown")
+		r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "FailedUpdate", "Failed to set autoscaling from zero annotations, instance type unknown or unavailable")
+
 		return ctrl.Result{}, nil
 	}
 
+	updateMachineSetAnnotations(machineSet, sku)
+
+	return ctrl.Result{}, nil
+}
+
+func getSKU(r *Reconciler, machineSet *machinev1.MachineSet) (resourceskus.SKU, error) {
+	providerConfig, err := getproviderConfig(machineSet)
+	if err != nil {
+		return resourceskus.SKU{}, mapierrors.InvalidMachineConfiguration("failed to get providerConfig: %v", err)
+	}
+
+	params := actuators.MachineScopeParams{
+		Machine: &machinev1.Machine{
+			Spec: machineSet.Spec.Template.Spec,
+		},
+		CoreClient: r.Client,
+	}
+
+	machineScope, err := actuators.NewMachineScope(params)
+	if err != nil {
+		return resourceskus.SKU{}, err
+	}
+
+	if r.resourceSkus == nil {
+		r.resourceSkus = resourceskus.NewService(machineScope)
```

Comment:
Do we need to cache the client?

Reply:
This was to allow inserting fake service in tests. It being cached was a bug and I have implemented your solution of using builder function